arangod/RocksDBEngine/RocksDBEdgeIndex.cpp

```diff
@@ -236,39 +273,54 @@ class RocksDBEdgeIndexLookupIterator final : public IndexIterator {
     _builder.clear();
   }
 
-  /// internal retrieval loop
+  // internal retrieval loop
   template<typename F>
   inline bool nextImplementation(F&& cb, uint64_t limit) {
     TRI_ASSERT(_trx->state()->isRunning());
-#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
-    TRI_ASSERT(limit > 0);  // Someone called with limit == 0. Api broken
-#else
-    // Gracefully return in production code
-    // Nothing bad has happened
+    TRI_ASSERT(limit > 0);
     if (limit == 0) {
+      // gracefully return in production mode. nothing bad has happened
       return false;
     }
-#endif
 
-    while (limit > 0) {
-      while (_builderIterator.valid()) {
-        // We still have unreturned edges in out memory.
-        // Just plainly return those.
-        TRI_ASSERT(_builderIterator.value().isNumber());
-        LocalDocumentId docId{
-            _builderIterator.value().getNumericValue<uint64_t>()};
-        _builderIterator.next();
-        TRI_ASSERT(_builderIterator.valid());
-        // For now we store the complete opposite _from/_to value
-        TRI_ASSERT(_builderIterator.value().isString());
+    std::string const* cacheKeyCollection = nullptr;
+    std::string const* cacheValueCollection = nullptr;
+
+    fu2::unique_function<void()> handleSingleResult = [&, this]() {
+      TRI_ASSERT(_builderIterator.value().isNumber());
+      LocalDocumentId docId{
+          _builderIterator.value().getNumericValue<uint64_t>()};
+      _builderIterator.next();
+      TRI_ASSERT(_builderIterator.valid());
+      // For now we store the complete opposite _from/_to value
+      VPackSlice fromTo = _builderIterator.value();
+      TRI_ASSERT(fromTo.isString());
+
+      std::string_view v = fromTo.stringView();
+      TRI_ASSERT(!v.empty());
+      if (v.front() == '/') {
+        // prefix-compressed value
+        cacheValueCollection =
+            _index->cachedValueCollection(cacheValueCollection);
```

Comment:
I don't understand why this is needed.

If I see correct it will do load only for case when cacheValueCollection is nullptr.
But it's true only if it's first call for cacheValueCollection.
If it's possible I don't understand why is cache in _index always already filled (because return value is never null, at least in assert).


Anyway, why don't move this load from first handleSingleResult call to this line
```
std::string const* cacheValueCollection = nullptr;
```

Reply:
The idea was to do this only lazily, when it is actually needed to perform the atomic load.

Reply:
I mean maybe just do always single load is better?

It's single load in all cases compared to 0 load in one cases and 1 load + n conditions in other.

Probably compiler can optimize conditions if you remove type erasure.
And probably branch prediction is very good.

But it looks like too much "probably" for single load (per function at all, not per handleSingleResult) in other cases.

Is it really optimize something?

Maybe I missed something

Reply:
The values of `cacheKeyCollection` and `cacheValueCollection` are data-dependent.
For example, `cacheKeyCollection` will be populated with the collection name of the edge on the first lookup in the index.
We really don't want to call this if the from/to value is invalid, if we don't have the cache enabled etc. That is also a reason why it is called lazily.